Cover/cmds/core/ping

llogen opened the pull request:
Refactored ping.go and added tests for it.

Signed-off-by: llogen <christophlange95@googlemail.com>

rminnich commented on cmds/core/ping/ping.go:
```diff
@@ -160,32 +160,43 @@ func ping1(net6 bool, host string, i uint64, waitFor time.Duration) (string, err
+
+func main() {
+	flag.Parse()
+	if *version {
```
We need to think about the version option. Either all u-root tools should do this uniformly, or none should. 

history: u-root commands did not have versions in the first years because u-root was designed to ship as source code. Versions seems superfluous. 
More history: versions predate git by about 40 years. The question is, what should version return? The version of u-root from whence it came? The ref of the git tree from whence it was built (I actually like this one)? Or some compound of the two? I think we need to think this through a bit mnore.

rminnich commented on cmds/core/ping/ping.go:
```diff
@@ -160,32 +160,43 @@ func ping1(net6 bool, host string, i uint64, waitFor time.Duration) (string, err
+		os.Exit(0)
+	}
+	// options without parameters (right now just: -hV)
+	if flag.NArg() < 1 {
```
suggestion: for now, just contain the version in the help. I.e., if someone does a -V, it can print out the "version" message as well. Or was the -V done to be compatible with GNU ping? GNU compatibility was an explicit non-goal of this project. But maybe we need compatibility?

rminnich commented on cmds/core/ping/ping_test.go:
```diff
@@ -0,0 +1,232 @@
+	return 0, nil
+}
+
+func (M *MyConn) Close() error                       { return nil }
```
YEP! This is how it's done. Very nice. Type implements the Conn interface.

rminnich commented on cmds/core/ping/ping_test.go:
```diff
@@ -0,0 +1,232 @@
+}
+
+func (M *MyConn) Close() error                       { return nil }
+func (M *MyConn) LocalAddr() net.Addr                { return nil }
```
but returning a nil pointer? Maybe, for example, this should return a valid address, e.g. an ip4 with 127.1? I suppose it's fine for now however.

BTW, you may be starting to see the rule that interfaces should be small.

rminnich commented on cmds/core/ping/ping_test.go:
```diff
@@ -0,0 +1,232 @@
+	"time"
+)
+
+type MyConn struct {
```
You can simplify this a  bit if you do this
type myConn struct {
net.IPConn
testRun string
}

Then, the ONLY functions you need implement are Read, and Write. This makes life simpler, should the Conn interface ever change, you are insulated from that change by the embedded net.Conn

Since testRun probably should not be exported, and in fact MyConn should not be exported, they might as well be lower case.

llogen replied:
Thank you for that!

rjoleary commented on cmds/core/ping/ping.go:
```diff
@@ -160,32 +160,43 @@ func ping1(net6 bool, host string, i uint64, waitFor time.Duration) (string, err
 	// ping needs to run forever, except if '*iter' is not zero
 	waitFor := time.Duration(*wtf) * time.Millisecond
 	var i uint64
 	for i = 1; *iter == 0 || i <= *iter; i++ {
```
This is a weirdly written for-loop. Could it just be:

```
for i := uint64(0); i < *iter; i++ {
    msg, err := p.ping1(*net6, host, i+1, waitFor)
```

llogen commented:
> Update the commit message to just read:
> 
> ```
> ping: Refactor ping.go and added tests for it
> ```

Pushed changes regarding your reviews and changed the commit message.

rminnich commented:
coverage is getting harder ... darn
